src/System.Linq.Expressions/src/System/Linq/Expressions/Compiler/StackSpiller.Temps.cs

```diff
@@ -205,6 +207,63 @@ private void EnsureDone()
                 }
             }
 
+            private static bool ShouldSaveToTemp(Expression expression)
+            {
+                if (expression == null)
+                    return false;
+
+                // Some expressions have no side-effects and don't have to be
+                // stored into temporaries, e.g.
+                //
+                //     xs[0] = try { ... }
+                //           |
+                //           v
+                //        t0 = xs
+                //        t1 = 0            // <-- this is redundant
+                //        t2 = try { ... }
+                //    t0[t1] = t2
+                //           |
+                //           v
+                //        t0 = xs
+                //        t1 = try { ... }
+                //     t0[0] = t1
+
+                switch (expression.NodeType)
+                {
+                    // Emits ldnull, ldc, initobj, closure constant access, etc.
+                    case ExpressionType.Constant:
+                    case ExpressionType.Default:
+                        return false;
+
+                    // Emits calls to pure RuntimeOps methods with immutable arguments
+                    case ExpressionType.Quote:
+                    case ExpressionType.RuntimeVariables:
+                        return false;
+
+                    case ExpressionType.MemberAccess:
+                        var member = (MemberExpression)expression;
+                        var field = member.Member as FieldInfo;
+                        if (field != null)
+                        {
+                            // Emits ldc for the raw value of the field
+                            if (field.IsLiteral)
+                                return false;
+
+                            // For read-only fields we could save the receiver, but
+                            // that's more involved, so we'll just handle static fields
+                            if (field.IsInitOnly && field.IsStatic)
+                                return false;
```

Comment:
I noticed that Roslyn applies the same optimization here. However, I had one thought/question here. According to ECMA-335, section I.8.9.5:

>  Also, under BeforeFieldInit semantics, type initializers are allowed to be executed at or before first access to any static field of that type, at the discretion of the CLI. 

The specification talks about `first access` here, which could be as late as the `ldsfld` instruction. If the type initializer throws, the timing of the `TypeInitializationException` changes in the presence of this optimization. I believe that in practice the type initializer runs earlier, at the point of encountering the first reference to the type in a body of IL code to be JITted. (And of course one can argue that type initializers shall not throw, nor that one should depend on the timing of such exceptions, even if an optimization like this moves the effect to a *later* point in the code.)